src/tactic/itauto.lean

```diff
@@ -0,0 +1,486 @@
+def and_kind.cmp (p q : and_kind) : ordering :=
+by { cases p; cases q, exacts [eq, lt, lt, gt, eq, lt, gt, gt, eq] }
+
+/-- A comparator for propositions. (There should really be a derive handler for this.) -/
+def prop.cmp (p q : prop) : ordering :=
+begin
+  induction p with _ ap _ _ p₁ p₂ _ _ p₁ p₂ _ _ p₁ p₂ _ _ p₁ p₂ generalizing q; cases q,
+  case var var { exact cmp p q },
+  case true true { exact eq },
+  case false false { exact eq },
+  case and and : aq q₁ q₂ { exact (ap.cmp aq).or_else ((p₁ q₁).or_else (p₂ q₂)) },
+  case or or : q₁ q₂ { exact (p₁ q₁).or_else (p₂ q₂) },
+  case imp imp : q₁ q₂ { exact (p₁ q₁).or_else (p₂ q₂) },
+  exacts [lt, lt, lt, lt, lt,
+          gt, lt, lt, lt, lt,
+          gt, gt, lt, lt, lt,
+          gt, gt, gt, lt, lt,
+          gt, gt, gt, gt, lt,
+          gt, gt, gt, gt, gt]
+end
+
+instance : has_lt prop := ⟨λ p q, p.cmp q = lt⟩
+
+instance : decidable_rel (@has_lt.lt prop _) := λ _ _, ordering.decidable_eq _ _
+
+end
+
+/-- A reified inductive proof type for intuitionistic propositional logic. -/
+@[derive has_reflect]
+inductive proof
+-- (n: A) ⊢ A
+| hyp (n : name) : proof
+-- ⊢ ⊤
+| triv : proof
+-- (p: ⊥) ⊢ A
+| exfalso' (A : prop) (p : proof) : proof
+-- (p: (x: A) ⊢ B) ⊢ A → B
+| intro (x : name) (A : prop) (p : proof) : proof
+-- ak = and:  (p: A ∧ B) ⊢ A
+-- ak = iff:  (p: A ↔ B) ⊢ A → B
+-- ak = eq:  (p: A = B) ⊢ A → B
+| and_left (ak : and_kind) (p : proof) : proof
+-- ak = and:  (p: A ∧ B) ⊢ B
+-- ak = iff:  (p: A ↔ B) ⊢ B → A
+-- ak = eq:  (p: A = B) ⊢ B → A
+| and_right (ak : and_kind) (p : proof) : proof
+-- ak = and:  (p₁: A) (p₂: B) ⊢ A ∧ B
+-- ak = iff:  (p₁: A → B) (p₁: B → A) ⊢ A ↔ B
+-- ak = eq:  (p₁: A → B) (p₁: B → A) ⊢ A = B
+| and_intro (ak : and_kind) (p₁ p₂ : proof) : proof
+-- ak = and:  (p: A ∧ B → C) ⊢ A → B → C
+-- ak = iff:  (p: (A ↔ B) → C) ⊢ (A → B) → (B → A) → C
+-- ak = eq:  (p: (A = B) → C) ⊢ (A → B) → (B → A) → C
+| curry (ak : and_kind) (A B : prop) (p : proof) : proof
+-- This is a partial application of curry.
+-- ak = and:  (p: A ∧ B → C) (q : A) ⊢ B → C
+-- ak = iff:  (p: (A ↔ B) → C) (q: A → B) ⊢ (B → A) → C
+-- ak = eq:  (p: (A ↔ B) → C) (q: A → B) ⊢ (B → A) → C
+| curry₂ (ak : and_kind) (B : prop) (p q : proof) : proof
+-- (p: A → B) (q: A) ⊢ B
+| app' : proof → proof → proof
+-- (p: A ∨ B → C) |- A → C
+| or_imp_left (A B : prop) (p : proof) : proof
+-- (p: A ∨ B → C) |- B → C
+| or_imp_right (A B : prop) (p : proof) : proof
+-- (p: A) |- A ∨ B
+| or_inl (B : prop) (p : proof) : proof
+-- (p: B) |- A ∨ B
+| or_inr (A : prop) (p : proof) : proof
+-- (p: B) |- A ∨ B
+-- (p₁: A ∨ B) (p₂: (x: A) |- C) (p₃: (x: B) |- C) |- C
+| or_elim (A B : prop) (p₁ : proof) (x : name) (p₂ p₃ : proof) : proof
+-- The variable x here names the variable that will be used in the elaborated proof
+-- (p: ((x:A) → B) → C) |- B → C
+| imp_imp_simp (x : name) (A B : prop) (p : proof) : proof
```

Comment:
Otherwise `app` would need access to the name generator, which would make it more cumbersome to use. It's easier to just pre-generate a name and store it here.